### PR TITLE
Compare two documentation outputs

### LINKED ISSUE #1801: Testsuite to compare the output of 2 doxygen runs
In order to keep track of the modification of doxygen, we should be able to compare the output of two runs. WIth `HTML_TIMESTAMP=NO`, it is possible to compare the html page directly but a layout change will result in important diff. Maybe looking at the xml output to see what's get exported is a solution. What is important to check is what classes, functions, members, ... end up in the documention.

### PR DESCRIPTION
**Pull request by maxGimeno**
This PR fixes #1801 

## Summary of Changes
It changes some details in the doxygen comments to avoid differences in the output of different versions of doxygen and adds a bash script to generate and diff two different versions of CGAL documentation

## Release Management

* Affected package(s): Documentation, PMP, AABB_tree, Polygon
* Issue(s) solved (if any): #1801

### CODE REVIEW

File: Documentation/doc/resources/1.8.13/BaseDoxyfile.in [hunk 1]
```diff
@@ -931,7 +931,9 @@ EXCLUDE_SYMBOLS        = Tr \
                          Cb \
                          Fb \
                          K \
-                         Traits
+                         Traits \
+                         internal \
+                         internal::* \
```
Comment (sloriot) [inline Documentation/doc/resources/1.8.13/BaseDoxyfile.in]: extra `\` I guess

File: Documentation/doc/resources/1.8.13/BaseDoxyfile.in [hunk 2]
```diff
@@ -1960,7 +1962,7 @@ MAN_LINKS              = NO
 # captures the structure of the code including all documentation.
 # The default value is: NO.
 
-GENERATE_XML           = NO
+GENERATE_XML           = YES
```
Comment (sloriot) [inline Documentation/doc/resources/1.8.13/BaseDoxyfile.in]: I don't think it should be ON by default. You can add an option in the cmake to switch it on/off.

File: Documentation/doc/resources/1.8.4/BaseDoxyfile.in [hunk 3]
```diff
@@ -816,7 +816,8 @@ EXCLUDE_SYMBOLS        = Tr \
                          Cb \
                          Fb \
                          K \
-                         Traits
+                         Traits \
+                         internal
```
Comment (sloriot) [inline Documentation/doc/resources/1.8.4/BaseDoxyfile.in]: is `internal::*` really needed?

File: Documentation/doc/resources/1.8.4/BaseDoxyfile.in [hunk 4]
```diff
@@ -1572,7 +1573,7 @@ MAN_LINKS              = NO
 # generate an XML file that captures the structure of
 # the code including all documentation.
 
-GENERATE_XML           = NO
+GENERATE_XML           = YES
```
Comment (sloriot) [inline Documentation/doc/resources/1.8.4/BaseDoxyfile.in]: Same remark as above.

File: Documentation/doc/scripts/compare_testsuites.sh [hunk 5]
```diff
@@ -0,0 +1,37 @@
+#!/bin/bash
+#Path to the CGAL_Documentation_build_directory/doc_output
+PATH_TO_DOC="$1"
+#path to the repository containing the output of this script for the reference documentation.
+DOC_REF="$2"
+#output in a new directory
```
Comment (sloriot) [inline Documentation/doc/scripts/compare_testsuites.sh]: You should check the number of arguments provided to the script and that the directories exist. Also you should show some error messages in both cases.

File: Documentation/doc/scripts/compare_testsuites.sh [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+mkdir -p doc_data
+cd ./doc_data
+for dir in $PATH_TO_DOC/*
+do
+  OUTPUT=$(basename $dir)
+  python ../documentation_parser.py $dir/xml > ./"$OUTPUT.txt"
```
Comment (sloriot) [inline Documentation/doc/scripts/compare_testsuites.sh]: Maybe you should handle the case when the python script does not end correctly.

File: Documentation/doc/scripts/documentation_parser.py [hunk 7]
```diff
@@ -0,0 +1,179 @@
...
+from collections import defaultdict
+from sys import argv 
+import os.path as op
+
+def check_type(_in, args):
+  res = False;
+  for arg in args:
+    res = res or _in == arg
+  return res
+
```
Comment (sloriot) [inline Documentation/doc/scripts/documentation_parser.py]: Comments please...

File: Polygon/include/CGAL/Polygon_2/Polygon_2_edge_iterator.h [hunk 8]
```diff
@@ -29,9 +29,9 @@
 #include <CGAL/circulator.h>
 
 namespace CGAL {
-
+#ifndef DOXYGEN_RUNNING //to avoid conflicts
 template <class Traits_, class Container_> class Polygon_2;
-
+#ifndef
```
Comment (sloriot) [inline Polygon/include/CGAL/Polygon_2/Polygon_2_edge_iterator.h]: I guess you mean `#endif`

File: Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/fair.h [hunk 9]
```diff
@@ -67,7 +67,7 @@ namespace internal {
   based on solving a linear bi-Laplacian system with boundary constraints,
   described in \cgalCite{Botsch2008OnLinearVariational}.
   The optional parameter `fairing_continuity` gives the ability to control the tangential
-  continuity C<sup>\a n</sup> of the output mesh.
+  continuity C<sup>\a n </sup>of the output mesh.
```
Comment (sloriot) [inline Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/fair.h]: mathjax should be used here or simply remove the useless `\a`

File: Documentation/doc/scripts/documentation_parser.py [hunk 10]
```diff
@@ -0,0 +1,179 @@
+from pyquery import PyQuery as pq
+from collections import defaultdict
+from sys import argv 
+import os.path as op
+
+def check_type(_in, args):
+  res = False;
+  for arg in args:
+    res = res or _in == arg
```
Comment (sloriot) [inline Documentation/doc/scripts/documentation_parser.py]:
```
  if _in==arg:
    return True
return False
```

File: Documentation/doc/scripts/documentation_parser.py [hunk 11]
```diff
@@ -0,0 +1,179 @@
...
+  return res
+
+root_path=argv[1]
+d = pq(filename=root_path+'/index.xml', parser="xml")
```
Comment (sloriot) [inline Documentation/doc/scripts/documentation_parser.py]: use `path.join` instead of +'/'

**Comment by lrineau** [comment]
@sloriot @maxGimeno Who is waiting for what, in this PR?

File: Documentation/doc/resources/1.8.13/BaseDoxyfile.in [hunk 12]
```diff
@@ -487,7 +487,7 @@ LOOKUP_CACHE_SIZE      = 0
 # normally produced when WARNINGS is set to YES.
 # The default value is: NO.
 
-EXTRACT_ALL            = YES
+EXTRACT_ALL            =
```
Comment (sloriot) [inline Documentation/doc/resources/1.8.13/BaseDoxyfile.in]: Why is this changed?

File: Documentation/doc/resources/1.8.4/BaseDoxyfile.in [hunk 13]
```diff
@@ -444,7 +444,7 @@ LOOKUP_CACHE_SIZE      = 0
 # Private class members and static file members will be hidden unless
 # the EXTRACT_PRIVATE respectively EXTRACT_STATIC tags are set to YES
 
-EXTRACT_ALL            = YES
+EXTRACT_ALL            =
```
Comment (sloriot) [inline Documentation/doc/resources/1.8.4/BaseDoxyfile.in]: why is this modified?

**Comment by sloriot** [comment]
Some of my comments were not answered. I added one more.

**Comment by sloriot** [comment]
I guess before integrating this, we need to see the output of a diff in a testsuite.

**Comment by sloriot** [comment]
`GENERATE_XML` should be set through a cmake variable otherwise I do not see how a testsuite could set it.

**Comment by lrineau** [comment]
Let's say it again: @sloriot @maxGimeno Who is waiting for what, in this PR? Maxime, you put this PR in the column "waiting for something". Can you tell us why?

**Comment by maxGimeno** [comment]
@sloriot  said "I guess before integrating this, we need to see the output of a diff in a testsuite.", so I thought we had to wait for that.

**Comment by lrineau** [comment]
Then can you please discuss about how to achieve that goal, instead of letting this PR stall?!

**Comment by maxGimeno** [comment]
[Here](https://cgal.geometryfactory.com/~mgimeno/Doc_results/result.html) is a link to the results of the tests.

**Comment by maxGimeno** [comment]
The tests are performed on cgal.geometryfactory.com everyday, using the maxGimeno/CGAL master branch. I voluntarily reverted the fixes done in Doxygen_compare-GF to have a good example of what a difference between versions looks like in the output of the script.

**Comment by lrineau** [comment]
@sloriot Is that the sort of result page you were expecting (see the issue #1801)?

**Comment by sloriot** [comment]
Looking at the page it is not clear what we are looking at:
- We have no information about which version of CGAL it is about
- We don't know what the diff is about (is it a run of doxygen master vs 1.8.4 patched or with another CGAL release?)
Is the page going to be updated each day? Past day results will be overwritten or are the results of each day appended ? How is that page will be accessible? Will the testsuite page be updated to have a link to that page?

**Comment by maxGimeno** [comment]
Ok, so I integrated your two points, you can check the new results page. From what we discussed, or at least what I remember, the page is supposed to be updated each day, the results would be overwritten and the page would be accessible through a link on the documentation testsuite page.

**Comment by lrineau** [comment]
The page give a 404 error, for now: 
  * https://cgal.geometryfactory.com/~mgimeno/Doc_results/result.html

The parent directory seems more interesting:
  * https://cgal.geometryfactory.com/~mgimeno/Doc_results/
but it shows an empty "diff" column.

**Comment by maxGimeno** [comment]
Yes it was in progress. The interesting link is now [this one](https://cgal.geometryfactory.com/~mgimeno/Doc_results/index.html). It shows the overview page for the documentation testsuite, with a new column containing the result of the diff. I had a hard time setting it correctly, that is why 2 lines are not correct.

**Comment by sloriot** [comment]
You might want to try https://github.com/amouat/diffxml
Also you should try setting `ALPHABETICAL_INDEX` to `YES` to see if we should add this option.

**Comment by maxGimeno** [comment]
diffxml doesn't work. Only applying diffxml from a 700 lines and a 1500 lines files generates a 4000 lines xml file, and applying patch to it like in the example leads to an error.

**Comment by lrineau** [comment]
I do not understand anything from https://cgal.geometryfactory.com/~mgimeno/Doc_results/, compared with the original result page https://cgal.geometryfactory.com/CGAL/Manual_doxygen_test/.

- What is the first column? How to know which internal release has been tested?
- I also do not know what is the "diff"? Diff between what and what? It should be written as an explanation at the top of the page.

**Comment by maxGimeno** [comment]
- The first column is the same than the one from Manual_doxygen_test, it is just the call I make to testsuite.py that differs as I don't have a CGAL_version to give. 
- Diff will be renamed in Diff with doxygen master when the index.html is recreated. For now it keeps using the existing one and thus do not apply my change to the column name.

**Comment by maxGimeno** [comment]
@sloriot Does it do what you wanted ? Is there still things to change ?

File: Documentation/doc/resources/1.8.13/BaseDoxyfile.in [hunk 14]
```diff
@@ -1960,7 +1961,7 @@ MAN_LINKS              = NO
 # captures the structure of the code including all documentation.
 # The default value is: NO.
 
-GENERATE_XML           = NO
```
Comment (sloriot) [inline Documentation/doc/resources/1.8.13/BaseDoxyfile.in]: no need to modify this line

File: Documentation/doc/resources/1.8.4/BaseDoxyfile.in [hunk 15]
```diff
@@ -1572,7 +1573,7 @@ MAN_LINKS              = NO
 # generate an XML file that captures the structure of
 # the code including all documentation.
 
-GENERATE_XML           = NO
```
Comment (sloriot) [inline Documentation/doc/resources/1.8.4/BaseDoxyfile.in]: no need to modify this line

**Comment by sloriot** [comment]
@maxGimeno The result page seems to be gone.

**Comment by maxGimeno** [comment]
@sloriot I still have it : https://cgal.geometryfactory.com/~mgimeno/Doc_results/index.html

**Comment by sloriot** [comment]
Then I don't have any remark for now.

**Comment by lrineau** [comment]
I have remarks, because the overview page as it is now is not really usable:
 - The script should get a way to display the name of the internal release (such as `CGAL-4.10-Ic-146`) instead of the Git hash.
 - Currently there is one link to one version of the documentation. I think we should be able to see both versions, to compare pages manually.
 - The script should display what is the mean of the diff column (display something like "diff between docs compiled with Doxygen versions 1.8.4 and 1.8.12").

**Comment by lrineau** [comment]
Note that I already said almost that six weeks ago (see https://github.com/CGAL/cgal/pull/1852#issuecomment-283693856).

**Comment by maxGimeno** [comment]
Update : https://cgal.geometryfactory.com/~mgimeno/Doc_results/index.html

**Comment by sloriot** [comment]
There is an issue with the navtree in `https://cgal.geometryfactory.com/~mgimeno/Doc_results/CGAL-4.10/master/Manual/index.html`. I guess the wrong resource files were taken. Also you should indicate that master is doxygen master. Maybe you can indicate the reduced hash of the doxygen branch used.

**Comment by lrineau** [comment]
- In the [overview page], one cannot know which version of Doxygen was used to produce the "Warnings" and "Errors" columns. I am surprised to see 26 warnings, whereas there are zero warnings in the [current test results].
- In the page [Doxygen Manual Results for CGAL-4.10], a reader of that page should now which two versions of doxygen were used. The two links could be named "documentation compiled by Doxygen 1.8.4" and "documentation compiled with Doxygen master (hashtag)".

Actually, instead of the simple hash tag, for Doxygen master, you could use the output of
```
git describe --tags
```
For the current master, that gives `Release_1_8_13-21-gb91f466`. That way you have both the last release tag, and the hash tag of the commit: here we see that there have been 13 commits since the tag `Release_1_8_13`, and the short hash is `gb91f466`.

  [overview page]: https://cgal.geometryfactory.com/%7Emgimeno/Doc_results/index.html
  [current test results]: https://cgal.geometryfactory.com/CGAL/Manual_doxygen_test/
  [Doxygen Manual Results for CGAL-4.10]: https://cgal.geometryfactory.com/%7Emgimeno/Doc_results/CGAL-4.10/index.html

**Comment by maxGimeno** [comment]
@sloriot Is there a way to configure cmake for using a specified resource directory ? Because the default is 1.8.4 and master won't work with it, it needs 1.8.13.

**Comment by sloriot** [comment]
Simply change the rule for choosing the default. If the version is >=1.8.13 then take that one.

**Comment by maxGimeno** [comment]
Update : https://cgal.geometryfactory.com/~mgimeno/Doc_results/index.html

**Comment by sloriot** [comment]
`Link to this documentation built with Doxygen 1.8.4 Link to documentation compiled with Doxygen Master Release_1_8_13-21-gb91f466` : compiled -> built + add `<br>`

**Comment by lrineau** [comment]
Thanks for the modifications.

The next step is to find a way to integrate that in the script `Scripts/developer_scripts/run_doxygen_testsuite`.

**Comment by lrineau** [comment]
And also find out why your page reports 26 warnings, whereas the current test results show 0 warnings.

**Comment by maxGimeno** [comment]
I tested locally the version for the tarballs, and I got this for BGL_doc: 
>citelist:4: Warning: Unexpected new line character
Error: Problems running dot: exit code=127, command='dot', arguments='"/home/gimeno/Bureau/test/cgal_doc_build/CGAL-4.10-Ic-182/doc/scripts/doc_dir/doc_output/BGL/./inline_dotgraph_1.dot" -Tpng -o "/home/gimeno/Bureau/test/cgal_doc_build/CGAL-4.10-Ic-182/doc/scripts/doc_dir/doc_output/BGL/dot_inline_dotgraph_1.png"'
Error: Problems running dot: exit code=127, command='dot', arguments='"/home/gimeno/Bureau/test/cgal_doc_build/CGAL-4.10-Ic-182/doc/scripts/doc_dir/doc_output/BGL/./inline_dotgraph_1.dot" -Tcmapx -o "/home/gimeno/Bureau/test/cgal_doc_build/CGAL-4.10-Ic-182/doc/scripts/doc_dir/doc_output/BGL/dot_inline_dotgraph_1.map"'

I have no idea what inline_dotgraph_1.dot is, anyone knows ?

**Comment by sloriot** [comment]
The `dot` program is used to draw [`this graph`](http://doc.cgal.org/latest/BGL/group__PkgBGLConcepts.html).

**Comment by maxGimeno** [comment]
Ok, thanks ! The graphviz package was missing on my machine.

**Comment by maxGimeno** [comment]
Update : https://cgal.geometryfactory.com/~mgimeno/Doc_results/index.html
Here is what it should look like when generated from run_doxygen_testsuite.

**Comment by lrineau** [comment]
Whouah! 👏 That's a lot better!

About the diff, I still do not get it. The [diff] shows:
```
diff -u -N -w ./doc_data/AABB_tree.txt /home/mgimeno/public_html/cgal-testsuite/cgal_doc_build/CGAL-4.10-Ic-183/doc/scripts/doc_ref/AABB_tree.txt
--- ./doc_data/AABB_tree.txt	2017-04-19 10:18:35.430722530 +0200
+++ /home/mgimeno/public_html/cgal-testsuite/cgal_doc_build/CGAL-4.10-Ic-183/doc/scripts/doc_ref/AABB_tree.txt	2017-04-19 10:14:59.016020196 +0200
@@ -355,6 +355,7 @@
       AABB_tree
       insert
       rebuild
+      number_of_intersected_primitives
       any_intersection
       any_intersected_primitive
       set_primitive_data_impl
```
but the function `AABB_tree::number_of_intersected_primitives` is however showed in both versions of the documentation [here with Doxygen-1.8.4][1.8.4] and [here with Doxygen 1.8.14][1.8.14]. So what is the meaning of that diff?

  [diff]: https://cgal.geometryfactory.com/~mgimeno/Doc_results/CGAL-4.10-Ic-183%20with%20Doxygen%201.8.4/diff.txt
  [1.8.4]: https://cgal.geometryfactory.com/~mgimeno/Doc_results/CGAL-4.10-Ic-183%20with%20Doxygen%201.8.4/output/AABB_tree/classCGAL_1_1AABB__tree.html#a91d75aedeed4ca4a4d44742d75c94f80
  [1.8.14]: https://cgal.geometryfactory.com/~mgimeno/Doc_results/CGAL-4.10-Ic-183%20with%20Doxygen%201.8.4/master/AABB_tree/classCGAL_1_1AABB__tree.html#a61658bd5a2ac1576bed456e35da71134

**Comment by maxGimeno** [comment]
The functions in the two versions are not of the same type. It is one of the few fixes that I made in this branch, but the testsuite is ran for integration, so my fixes are not taken into account.

**Comment by lrineau** [comment]
@maxGimeno @sloriot 

Do you have an idea what are the transition steps? Should we first merge the PR (that modified some CGAL headers)? Or should we first switch the version of the script that generate the documentation, and then merge the PR?

**Comment by maxGimeno** [comment]
I thought everything was in the PR. I mean if you merge the PR, the script
in developper_scripts will be updated anyway, won't it ?

2017-05-30 14:33 GMT+02:00 Laurent Rineau <notifications@github.com>:

**Comment by lrineau** [comment]
Yes, that is right. But I have at least a manual step to do on my $HOME on the server, to update the script.

One more question: how will the new script deal with the existing HTML page in `/srv/CGAL/www/Members/Manual_doxygen_test`. Will it just happen a new line, with the new format? But the old lines and the new lines do not have the same number of columns.

**Comment by maxGimeno** [comment]
I would first merge the PR, as the results fetched by your script should not be altered, or I hope so.

For your other question, I'm afraid we'll have to make a clean start, as the python script does not update an existing header, it will only change if a new one is created.

**Comment by lrineau** [comment]
@sloriot How would do for the transition?

**Comment by sloriot** [comment]
Can we use different URL for the upload and keep both versions ? (at least for 10 internal releases)

**Comment by lrineau** [comment]
The merge of this PR in CGAL-4.11-Ic-30 is the reason why there was no Doxygen testsuite last night. There is a problem...

**Comment by maxGimeno** [comment]
I think the problem is that in testsuite.py I added an argument 'master-dir" and did not anticipate the case where it is not provided.

**Comment by lrineau** [comment]
It seems that worked, tonight!

https://cgal.geometryfactory.com/CGAL/Manual_doxygen_test/

**Comment by maxGimeno** [comment]
Great !

**Comment by lrineau** [comment]
For @sloriot: it seems that just merging the PR in `integration` changes the scripts that are used by our cron job.

Fortunately, the new script is able to add a new row of results in the table, even if the headers row is not the right one. I will just have to change the headers row by hand.
 
The downside is that I do not see any easy way to run both versions of the script in parallel, to compare.

**Comment by sloriot** [comment]
@maxGimeno  It seems that the post processing is not done with Doxygen 1.8.4

**Comment by maxGimeno** [comment]
Post processing ?

**Comment by lrineau** [comment]
I agree. It seems that, in the new scripts, only `make doc` is used to compile the documentation, but not `make doc_with_postprocessing`.

The old script used `make doc_and_publish_testsuite`, that depends on the target `doc_with_postprocessing`.
